sys/include/doip.h

```diff
@@ -0,0 +1,180 @@
+#define UDS_RESPONSE_CODES_CNC      0x22
+#define UDS_RESPONSE_CODES_RSE      0x24
+#define UDS_RESPONSE_CODES_NRFSC    0x25
+#define UDS_RESPONSE_CODES_FPEORA   0x26
+#define UDS_RESPONSE_CODES_ROOR     0x31
+#define UDS_RESPONSE_CODES_SAD      0x33
+#define UDS_RESPONSE_CODES_IK       0x35
+#define UDS_RESPONSE_CODES_ENOA     0x36
+#define UDS_RESPONSE_CODES_RTDNE    0x37
+#define UDS_RESPONSE_CODES_UDNA     0x70
+#define UDS_RESPONSE_CODES_TDS      0x71
+#define UDS_RESPONSE_CODES_GPF      0x72
+#define UDS_RESPONSE_CODES_WBSC     0x73
+#define UDS_RESPONSE_CODES_RCRRP    0x78
+#define UDS_RESPONSE_CODES_SFNSIAS  0x7E
+#define UDS_RESPONSE_CODES_SNSIAS   0x7F
+
+#define UDS_SUPRESS_REPLY           0x80
+
+#define DOIP_GENERIC_DOIP_HEADER_NEGATIVE_ACK           0x0000
+#define DOIP_VEHICLE_IDENTIFICATION_REQUEST             0x0001
+#define DOIP_VEHICLE_IDENTIFICATION_REQUEST_WITH_EID    0x0002
+#define DOIP_VEHICLE_IDENTIFICATION_REQUEST_WITH_VIN    0x0003
+#define DOIP_VEHICLE_IDENTIFCATION_RESPONSE             0x0004
+#define DOIP_ROUTING_ACTIVAION_REQUEST                  0x0005
+#define DOIP_ALIVE_CHECHK_REQUEST                       0x0007
+#define DOIP_ALIVE_CHECK_RESPONSE                       0x0008
+#define DOIP_DOIP_ENTITY_STATUS_REQUEST                 0x4001
+#define DOIP_DOIP_ENTITY_STATUS_RESPONSE                0x4002
+#define DOIP_DIAGNOSTIC_POWER_MODE_INFO_REQUEST         0x4003
+#define DOIP_DIAGNOSTIC_POWER_MODE_INFO_RESPONSE        0x4004
+#define DOIP_DIAGNOSTIC_MESSAGE                         0x8001
+#define DOIP_DIAGNOSTIC_MESSAGE_POSITIVE_ACK            0x8002
+#define DOIP_DIAGNOSTIC_MESSAGE_NEGATVE_ACK             0x8003
+/* ########################*/
+
+
+/* ##### DoIP Types #### */
+
+/**
+ * @brief doip logical sa address
+ */
+typedef uint16_t doip_sa;
+
+/**
+ * @brief doip logical ta address
+ */
+typedef uint16_t doip_ta;
+
+/**
+ * @brief doip logical ta address type
+ *
+ * @detailed Extension to the doip_ta parameter.
+ *      It is used to encode the communication model used by the communicating
+ *      peer entities of the doip layer.  Two communication models are specified:
+ *      1 to 1 communication, called physical addressing (unicast), and 1 to n
+ *      communication, called functional addressing (multicast/broadcast).
+ */
+typedef enum doip_tat { physical, functional, doip_tat_MAX } doip_tat;
+
+/**
+ * @brief Outcome of service execution
+ *
+ * @detailed Outcome of service execution. If two or more errors are discoverd at the same time, the parameter first found in this list will be given
+ */
+typedef enum doip_result {
+    DOIP_OK,
+    DOIP_HDR_ERROR,
+    DOIP_TIMEOUT_A,
+    DOIP_UNKNOWN_SA,
+    DOIP_INVALID_SA,
+    DOIP_UNKNOWN_TA,
+    DOIP_MESSAGE_TOO_LARGE,
+    DOIP_OUT_OF_MEMORY,
+    DOIP_taRGET_UNREACHABLE,
+    DOIP_NO_LINK,
+    DOIP_NO_SOCKET,
+    DOIP_ERROR
+} doip_result;
+
+
+
+/* ####################*/
+
+
+/* ####### DoIP Functions ########## */
```

Comment:
I miss a `sock_doip_t`here, which holds the underlying udp socket, tcp socket, static buffers and other state. This socket can then be created by a `sock_doip_create` function and closed by a `sock_doip_close`function, similar to the other protocols which implement a socket interface. See eg. `sys/include/net/sock/udp.h`

The biggest benefit of this is that the socket contains all the state of a connection, allowing to open multiple connections.

Reply:
Should the code be POSIX compliant? Because if I'm not mistaken, POSIX has _t reserved for their own structs.

Reply:
it's fine to use _t here.